### ISSUE TITLE
UI theme improvement

### ISSUE THREAD

**Issue by nicolas-p** [slavapestov/factor#1389]
I am proposing a small UI theme redesign which I think is a nice improvement.

This is a screenshot of the current UI:

![current factor ui](https://cloud.githubusercontent.com/assets/11431580/8683274/07bdf6e8-2a72-11e5-965f-edc7b501ad59.png)

This is a mock-up I made:

![redesigned factor ui](https://cloud.githubusercontent.com/assets/11431580/8683286/14c6605a-2a72-11e5-9605-55f113c2ba96.png)

List of changes:
- top buttons unified in a toolbar (colour F2F2F2 with a 1-px border of colour B3B3B3)
- "Search" label placed inside the search box and replaced by the input when the user types
- headers extend the full width of the window
- some text placement corrections
- Apple-style scroll bars: hidden when not needed, simple black bars otherwise
- status bar height increased by 5 px
- tip of the day removed (or at least there should be a cross to close it; in this case the blue background should also extend the full width of the window)

What do you think?

**Comment by mrjbq7** [comment]
That looks nice, we implement our own UI in OpenGL so building "apple-style" scroll-bars might be a little harder than changing the theme but I'd have to look into it more.

I particularly like your button theme, much more "modern".

The headers extended full width of the window looks really great.

I like seeing the tip-of-the-day on new listener windows, do you find it annoying?

Have you had a chance to implement any of these, or is this just a mockup at this point?

Thanks!

**Comment by nicolas-p** [comment]
I'm glad you like it!

Concerning scrolI-bars, I have just obtained something close to what I had in mind just by modifying the TIFF images and restarting Factor. I end up with a lot of transparent images.

I shall try to implement the other ideas in the next few days.

I've never liked tips of the day in any software. But here they're clearly not as intrusive as the pop-up windows we see often.

**Comment by mrjbq7** [comment]
I generally agree on tips of the day, but figured most people are unaware of all the features Factor provides so thats why I'm a little biased the other way. :-)

**Comment by nicolas-p** [comment]
There's progress. It is not a mock-up this time, this is a real screenshot:

![capture d ecran 2015-07-16 a 22 03 07](https://cloud.githubusercontent.com/assets/11431580/8733882/81035242-2c06-11e5-839c-20eca7f268f3.png)


Implemented:
- scroll-bars
- unified toolbar (except border)
- search label removed
- documentation header extending the full width and always visible at the top when scrolling
- documentation "previous"/"next" links moved at the bottom of each page (they were taking too much space in my fixed header, and I figured they are mostly useful when you reach the bottom)
- tip of the day removed (but don't worry, I won't include that in my future pull request)

I am able to draw a border on the unified toolbar, but it predictably displays the four borders when I only want the bottom one. I don't know if there's a simple solution.

As I said the other day, I wanted to have the "Search" label placed inside the search box, and replaced by the text when typing. I don't know how difficult this might be to implement, but I think everybody can guess this is a search box anyway.

**Comment by mrjbq7** [comment]
That looks pretty neat!  Do you have a diff or a set of commits I can see what you've changed?

You might want to set the search field to be white instead of background colored:

```diff
--- a/basis/ui/tools/browser/browser.factor
+++ b/basis/ui/tools/browser/browser.factor

 : <search-field> ( browser -- field )
     '[ _ search-browser ] <action-field>
         10 >>min-cols
-        10 >>max-cols ;
+        10 >>max-cols
+        COLOR: white <solid> >>interior ;
```

**Comment by nicolas-p** [comment]
If you execute the following code in the Listener, you should have the same thing as me except for the TIFF images. This code is probably a bit dirty, this should be considered a draft. I shall create proper git commits later.

```factor
IN: ui.gadgets.buttons.private
CONSTANT: button-background COLOR: grey95

IN: ui.gadgets.buttons
: add-toolbar ( track -- track )
    dup <toolbar> { 3 3 } <border> 
    button-background <solid> >>interior
  ! COLOR: gray <solid> >>boundary 
    align-left f track-add ;

IN: help.tips
: tip-of-the-day. ( -- ) ;

IN: ui.tools.browser
: <browser-toolbar> ( browser -- toolbar )
    horizontal <track>
        0 >>fill
        1/2 >>align
        { 5 5 } >>gap
        over <toolbar> f track-add
        swap search-field>> 1 track-add ;

IN: ui.tools.browser       
: <browser-gadget> ( link -- gadget )
    vertical browser-gadget new-track
        1 >>fill
        swap >link <model> >>model
        dup <history> >>history
        dup <search-field> >>search-field
        dup <browser-toolbar> { 3 3 } <border> 
        button-background <solid> >>interior 
        { 1 0 } >>fill f track-add
        dup dup <help-pane> { 10 0 } <border> { 1 1 } >>fill
        <scroller> >>scroller scroller>> 1 track-add ;
        
        
IN: ui.tools.browser    
: <help-title> ( browser-gadget -- gadget )
    model>> [ '[ _ $title ] try ] <pane-control> ;

IN: ui.tools.browser       
: <browser-gadget> ( link -- gadget )
    vertical browser-gadget new-track
        1 >>fill
        swap >link <model> >>model
        dup <history> >>history
        dup <search-field> >>search-field
        dup <browser-toolbar> { 3 3 } <border> 
          button-background <solid> >>interior 
          { 1 0 } >>fill f track-add
        dup <help-title> { 3 3 } <border>
          FactorLightTan <solid> >>interior
          { 1 0 } >>fill f track-add
        dup dup <help-pane> { 10 0 } <border> { 1 1 } >>fill
        <scroller> >>scroller scroller>> 1 track-add ;

IN: help
: ($navigation1) ( topic -- )
    help-path-style get
       [ help-path [ reverse $breadcrumbs ] unless-empty ]
    with-style ;

IN: help    
: ($navigation2) ( topic -- )
    help-path-style get 
       [ $navigation-table ]
    with-style ;
    
IN: help        
: print-topic ( topic -- )
    >link
    last-element off
    [ nl article-content print-content nl ] 
    [ ($blank-line) ($navigation2) ] 
    bi ;
    
IN: help     
: $title ( topic -- )
    title-style get [
        title-style get [
            [ ($title) ] [ ($navigation1) ] bi
        ] with-nesting
    ] with-style ;

IN: ui.tools.browser       
: <search-field> ( browser -- field )
    '[ _ search-browser ] <action-field>
        10 >>min-cols
        10 >>max-cols
        COLOR: white <solid> >>interior ;
```

**Comment by mrjbq7** [comment]
This is great!  Now I can see how I might progressively apply some of these changes.

How do you like the toolbar style in popup windows... for example, type "asdf" at the listener and hit Enter, you get an error popup.

**Comment by nicolas-p** [comment]
The new toolbar style is automatically applied everywhere. I had already looked at error popups, traceback and inspector windows and I think it looks good everywhere.

![capture d ecran 2015-07-17 a 07 36 36](https://cloud.githubusercontent.com/assets/11431580/8741282/9b0cb0de-2c56-11e5-941b-90957e7bd641.png)

**Comment by nicolas-p** [comment]
I finally managed to draw a 1-px line under the toolbar and under the documentation header. It looks really good. But it is really a hack. I created a pile with a grey background with two shelves inside and I set the gap to 1 px.

```factor
IN: ui.gadgets.buttons
: add-line ( track -- track )
    <pile> { 0 1 } >>gap COLOR: grey70 <solid> >>interior 
    <shelf> add-gadget
    <shelf> add-gadget
    f track-add ;
```

![capture d ecran 2015-07-17 a 18 32 33](https://cloud.githubusercontent.com/assets/11431580/8752112/5a56fba6-2cb2-11e5-83c9-e5f6993bf7d8.png)

I've updated the entire code that I posted yesterday with the new changes.

**Comment by mrjbq7** [comment]
Thats a nice hack, there ought to be a straightforward way to make a ``pen`` that when you call ``draw-boundary`` only draws the bottom border.

**Comment by nicolas-p** [comment]
I have this solution:
```factor
IN: ui.pens.solid
TUPLE: solid-underlined < solid ;

: <solid-underlined> ( color -- solid-underlined ) 
    solid-underlined new swap >>color ;
    
: bottom-verticies-only ( vertices -- bottom-vertices )
  { 6 7 4 5 4 5 6 7 6 7 } swap nths ;
  
M: solid-underlined draw-boundary
    [ (solid) ] [ boundary-vertices>> bottom-verticies-only gl-vertex-pointer ] bi
    (gl-rect) ;
```
But it would be better to redefine directly the ```boundary-verticies>>``` accessor of the ```solid-underlined``` class. For this I need to call the ```boundary-verticies>>``` accessor of the parent class. Is it possible to do that?

**Comment by mrjbq7** [comment]
```factor
M: solid-underlined boundary-vertices>>
    call-next-method bottom-vertices-only ;
```

**Comment by mrjbq7** [comment]
For what it's worth, it might be better to compute the vertices less often:

```factor
TUPLE: solid-underlined < solid ;

: <solid-underlined> ( color -- solid-underlined )
    solid-underlined new swap >>color ;

M: solid-underlined recompute-pen
    [ call-next-method ] keep [
        { 6 7 4 5 4 5 6 7 6 7 } swap nths
    ] change-boundary-vertices drop ;
```

(I understand you'll likely go a different direction with 1px gaps in a layout that draws solid background, but for perpetuity I'm showing another way here).

**Comment by mrjbq7** [comment]
Merged #1396, closing this!
